### PR TITLE
fix: incoming call keeps showing up even after rejecting it (AR-2934)

### PR DESCRIPTION
**Pull request by ohassine**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

Call keeps popping up many times even after rejecting it.

 STEPS TO REPRODUCE
- Login with the same account on two different devices ( one of them AR)
- When you receive a call, reject or answer it on the other platform, not Reloaded
- You should see the incoming call screen appearing multiple times (randomly) 
You will be able to reproduce it on old accounts, not fresh installs

### Causes (Optional)

The root cause is that call status is not changed in database and stays on INCOMING state.

#### Why?
When we receive a call we create a call object with `INCOMING` state and we save it on database. If the call got answered reject on other platform then we update the status to `CLOSED`.

On old accounts, due some performance issues, the flow is inverted:
 - The call to update is invoked first(You already have some closed calls with that person) through `OnCloseCall` callback, so it passes
 - Then create call object is called next with `INCOMING` state through `OnIncomingCall` callback

Since `OnCloseCall` already happened and there is the only place where we update the status to CLOSED
--> we have now  a call saved in database with `INCOMING` state that will never change. And at this time incoming screen will start popping up ⬆️ 

### Solutions

This PR will not fix the issue of inverted calls where `UPDATE` called before `CREATE`. This needs some refactoring.

In this PR I added a new status called `CLOSED_INTERNALLY`  that will be used when the user terminates the ongoing call.
Also, I used the REJECTED status when the user rejects an incoming call.

This way even that issue happens, we have a way to stop it.

#### Test Coverage (Optional)

- [ ] I have added automated test to this contribution

#### How to Test

Steps Above 👆 

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/call/usecase/RejectCallUseCase.kt [hunk 1]
```diff
@@ -1,14 +1,27 @@
...
-class RejectCallUseCase(private val callManager: Lazy<CallManager>) {
+class RejectCallUseCase(
+    private val callManager: Lazy<CallManager>,
+    private val callRepository: CallRepository,
+    private val dispatchers: KaliumDispatcher = KaliumDispatcherImpl
+) {
+
+    suspend operator fun invoke(conversationId: ConversationId) = withContext(dispatchers.io) {
+        callingLogger.d("[RejectCallUseCase] -> Updating call status to CLOSED_INTERNALLY")
+        callRepository.updateCallStatusById(conversationId.toString(), CallStatus.REJECTED)
```
Comment (alexandreferris) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/call/usecase/RejectCallUseCase.kt:23]:
```suggestion
        callingLogger.d("[RejectCallUseCase] -> Updating call status to REJECTED")
        callRepository.updateCallStatusById(conversationId.toString(), CallStatus.REJECTED)
```

File: logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/call/EndCallUseCaseTest.kt [hunk 2]
```diff
@@ -55,5 +64,65 @@ class EndCallUseCaseTest {
...
+
+        verify(callRepository)
+            .function(callRepository::updateIsCameraOnById)
+            .with(eq(conversationId.toString()), eq(false))
+            .wasInvoked(once)
+
+        verify(callRepository)
+            .suspendFunction(callRepository::persistMissedCall)
+            .with(any())
+            .wasNotInvoked()
```
Comment (alexandreferris) [inline logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/call/EndCallUseCaseTest.kt:110]: Shouldn't this be invoked per name of the test? 🤔

File: logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/call/EndCallUseCaseTest.kt [hunk 3]
```diff
@@ -55,5 +64,65 @@ class EndCallUseCaseTest {
...
+    @Test
+    fun givenNonEstablishedCall_whenEndCallIsInvoked_thenSaveMissedCall() = runTest {
+        given(callRepository)
+            .suspendFunction(callRepository::persistMissedCall)
+            .whenInvokedWith(eq(conversationId))
+            .thenDoNothing()
+
+        given(callRepository)
+            .suspendFunction(callRepository::establishedCallsFlow)
+            .whenInvoked().then { flowOf(listOf(call.copy(status = CallStatus.INCOMING))) }
```
Comment (alexandreferris) [inline logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/call/EndCallUseCaseTest.kt:83]: `establishedCallsFlow` returns only calls with `ESTABLISHED` or `ANSWERED` status, for incoming calls we should use `::incomingCallsFlow`

File: logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/call/RejectCallUseCaseTest.kt [hunk 4]
```diff
@@ -36,12 +41,22 @@ class RejectCallUseCaseTest {
...
         rejectCallUseCase.invoke(conversationId)
 
         verify(callManager)
             .suspendFunction(callManager::rejectCall)
             .with(eq(conversationId))
             .wasInvoked(once)
+
+        verify(callRepository)
+            .suspendFunction(callRepository::updateCallStatusById)
+            .with(eq(conversationId.toString()),  eq(CallStatus.REJECTED))
```
Comment (alexandreferris) [inline logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/call/RejectCallUseCaseTest.kt:58]:
```suggestion
            .with(eq(conversationId.toString()), eq(CallStatus.REJECTED))
```

Just some unnecessary extra space here

**Review by alexandreferris — COMMENTED** [review]
Great work addressing this issue, I do have some comments so let me know if something doesn't make sense

File: logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/call/EndCallUseCaseTest.kt (hunk 3 above)
Comment (ohassine) [reply logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/call/EndCallUseCaseTest.kt:83]: Yeah right . I will update it to return empty list

File: logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/call/EndCallUseCaseTest.kt (hunk 2 above)
Comment (alexandreferris) [reply logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/call/EndCallUseCaseTest.kt:110]: the name of the test is: `givenNonEstablishedCall_whenEndCallIsInvoked_thenSaveMissedCall`, so shouldn't `callRepository::persistMissedCall` be invoked at least once here?

Comment (ohassine) [reply logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/call/EndCallUseCaseTest.kt:110]: Oh! yes good catch

**Review by alexandreferris — APPROVED** [review]
Great work here!

Something we need to keep in mind now (doesn't need to be done in this PR), but to update:

`updateLastCallStatusByConversationId` and `updateOpenCallsToClosedStatus` to maybe add the status `REJECTED`.

**Pull request merged**